### ISSUE TITLE
iTunes / Spotify Integration

### ISSUE THREAD

**Issue by blahsd** [blahsd/snwe#6]

**Comment by paanvaannd** [comment]
The iTunes module is not working for me.

After opening the preference window and selecting the iTunes option, opening iTunes, and playing a song, there is no song information shown. The player buttons are interactable but do not control iTunes. Opening the preference window again shows that the iTunes option is no longer selected and that no option is selected under the music player menu. It seems that the iTunes selection in that menu is not being persisted.

**Comment by blahsd** [comment]
Alright, this is going to be a tough one because I can't reproduce it, even after deleting everything and doing a clean clone of the repo. I've opened a new branch 'debug' in order to pin this down. If you clone from that branch, the following should happen:

- When you open the preference pane, you should get an alert indicating which module is loaded
- When you select a different option, you should get (a) two alerts, one from the preference pane and one from the bar itself, indicating the currently loaded module, and (b) two alerts, indicating the module that is getting loaded.

- Whenever you select any other preference you should get the same alerts, which are not useful in that case. 

If this happens, the module is getting loaded and we'll have to look somewhere else.
**If this doesn't happen** something is wrong with the loading of the module. 

Let me know!

**Comment by paanvaannd** [comment]
I'm on the most recent commit, switched to the debug branch, and ran the application again (after deleting the app/css/colors-wal.css file) but there's no change in behavior... I don't see any notifications nor can I see track info, unfortunately.

 I don't have Spotify installed. Is this set up similar to the supernerd widget so that it would work only if Spotify and iTunes are installed on the same system?

**Comment by blahsd** [comment]
> I don't have Spotify installed. Is this set up similar to the supernerd widget so that it would work only if Spotify and iTunes are installed on the same system?

Nope. It's two separate modules so it should be absolutely irrelevant. For good measure I uninstalled it from my system too, and it still works for me. 

If you want, you could try activating the debug window on your copy of snwe to help me pin down the problem. Here's what you'd have to do:

1. Navigate to the ../snwe folder
2. Open main.js. At line 24 you should have the following: 

`//  win.webContents.openDevTools();`

Remove the two slashes to uncomment the line and activate it. 

3. At line 14 you should have:

`height:     64,`

Substitute '64' with 'height' in order to have: 

`height:     height,`

4. Open ./js/windowManager.js. At line 39 you should have the following:

`//popWindow.webContents.openDevTools();`

Again, uncomment the line.

What you are doing is the following: 2 and 4 make it so that spawning a new window does so while also spawning the developer tools, which include a console that the program uses to pass output to the user. Step 3 is needed so that the console is actually large enough for you to see the output. 

In the developer tools, click on the tab named 'Console'. You should now see that most things you do (clicking around, changing preferences, etc.) result in some kind of message being printed (some might not be very helpful or sound stupid, they weren't intended to be seen by others :P). You might see red lines indicating errors. What you should do is try to open up the preference pane a couple of times, change preferences, close it down, etc. and copy here your console output. If you see anything suspicious (esp. red lines ofc) take a note of what you did that got you a red line, and see if you can reproduce it. 

Thanks!!!

**Comment by paanvaannd** [comment]
Made the above changes to the program files and after running `npm start` and opening the resulting console view, I get the following error immediately on startup:

<img width="552" alt="error" src="https://user-images.githubusercontent.com/16567134/39820786-9bc23cfe-5374-11e8-9c44-af05ce04e707.png">

_Username redacted, but the path is just the expansion if $HOME/bin/snwe/app/js/itunes_

If I'm interpreting this properly, it seems that it is not reading the itunes.js file properly. Strange, considering that file exists in the appropriate directory...

Here is the output of tree in the cloned repo (ignoring the node_modules and .git directories):

```
~/bin/snwe debug*
❯ tree -a -I '.git|node_modules'
.
├── .DS_Store
├── .gitignore
├── README.md
├── app
│   ├── css
│   │   ├── center.css
│   │   ├── colors-wal.css -> /Users/pavananand/.cache/wal/colors.css
│   │   ├── colors.css
│   │   ├── common.css
│   │   ├── flat.css
│   │   ├── four.css
│   │   ├── mono.css
│   │   └── settings.css
│   ├── index.html
│   ├── js
│   │   ├── index.js
│   │   ├── itunes.js
│   │   ├── mpd.js
│   │   ├── settings.js
│   │   ├── settingsManager.js
│   │   ├── spotify.js
│   │   └── windowManager.js
│   └── settings.html
├── log.txt
├── main.js
├── package-lock.json
└── package.json

3 directories, 24 files
```

**Comment by blahsd** [comment]
Alright I think I got it. The preference pane is looking for */itunes* instead of */itunes.js* by mistake.

I think you had your preferences set by a previous commit which wasn't setting them correctly, then I fixed it, but your preferences were fucked up now, so it couldn't even get to the point of loading them to fix them. Try this:

Open settingsManager.js – at the very end, make a new line and add the following: 

 `initializePreferences();` 

This will call upon the function that sets up fresh preferences – it's generally only used when the preferences are not defined yet, but in your case it will wipe your preferences (sorry) and set them up correctly. 
If I'm right this will fix the issue. After it's fixed, you should go back and delete that line, so that your preferences go back to being persistent. 

You're the best debugger I've ever had btw hahaha

**Comment by paanvaannd** [comment]
![its_working_star_wars](https://user-images.githubusercontent.com/16567134/39825282-7132ece4-537f-11e8-8e22-c81c93f7e5fd.gif)

Made the above change to settingsManager.js, started the debug branch app, opened iTunes, and started playing a song. Works as expected, now!

This does resolve issue #3 as well! Themes change now (and without needing to restart), and I can change from the pywal colorscheme to the defauls colorscheme as well (but not back to the pywal colorscheme without restarting). Regenerating a pywal theme while snwe is open no longer results in snwe disappearing.

Thank you! You're one of the best devs I've helped debug software for :+)

**Comment by paanvaannd** [comment]
One more thing I noticed:

<img width="1280" alt="error" src="https://user-images.githubusercontent.com/16567134/39828526-607d1a1e-5389-11e8-92c4-82ee92bc1e02.png">

It seems like there are errors thrown when certain osascript functions are executed? I register these errors in the preference window debugging window but not in the main interface window. The first error refers to the function in settingsManager.js that creates a symlink to the pywal-generated CSS file and the second error is the script in itunes.js that checks if iTunes is playing and, if it is, to get the artist and track name. The second errorr is mirrored when changing the Player option to Spotify since there are similar scripts in spotify.js.

So far as I know, the functions run just fine since everything appears to be working well. I'm not sure why those errors are there.
